tinygrad/helpers.py

```diff
@@ -99,3 +109,36 @@ class GlobalCounters:
   cache: ClassVar[Optional[List[Tuple[Callable, Any]]]] = None
   @staticmethod
   def reset(): GlobalCounters.global_ops, GlobalCounters.global_mem, GlobalCounters.time_sum_s, GlobalCounters.kernel_count, GlobalCounters.cache = 0,0,0.0,0,None
+
+# Stripped down version of a WeakSet
```

Comment:
Wow, are these classes really faster? If so, even though it adds lines, I'm supportive of this. It surfaces complexity.

Do we really need both a Set and a Dictionary, or is just a Dictionary enough?

Reply:
Yeah this is faster because the stdlib weak data structures are threadsafe-ish and iteration safe, both features we don't need.

I think we do need both because the dict is more expensive than the set so we only use the dict where it is absolutely necessary (for the addressable cache in `lazy.py`)

Reply:
I think these classes should probably have some unit tests.

Also, I propose moving them to `lazy.py`, that's the only place that should use this weakref stuff.